### PR TITLE
No particle field matrix

### LINKED ISSUE #1051: Make sure particle fields do not create a matrix block
to save memory. Do this by specifying Coupling::None for the SparsityPattern

### PR DESCRIPTION
**Pull request by class4kayaker**
Fixes #1051.

Includes a small test case intended to test correct matrix allocation behavior for varying composition methods.

### CODE REVIEW

File: source/simulator/core.cc [hunk 1]
```diff
@@ -1030,8 +1030,25 @@ namespace aspect
...
+                coupling[x.compositional_fields[c]][x.compositional_fields[c]]
+                  = DoFTools::always;
+                break;
+              // Non-coupling cases
+              case Parameters<dim>::AdvectionFieldMethod::particles:
+                coupling[x.compositional_fields[c]][x.compositional_fields[c]]
+                  = DoFTools::none;
+                break;
+              default:
+                AssertThrow(false,ExcNotImplemented());
```
Comment (bangerth) [inline source/simulator/core.cc]: Good style -- you've really been learning quickly!

Comment (bangerth) [inline source/simulator/core.cc]: Although I would probably have written this as a regular `Assert`.

File: tests/matrix_nonzeros_2/screen-output [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+   Postprocessing:
+                                   
+Total system matrix memory consumption: 0.04 MB.
+Total system matrix nnz: 1,090
+system matrix nnz by block: 
+         288         112           0           0           0
+         112           0           0           0           0
+           0           0         289           0           0
+           0           0           0         289           0
+           0           0           0           0           0
```
Comment (bangerth) [inline tests/matrix_nonzeros_2/screen-output]: nice test case!

File: source/simulator/core.cc (hunk 1 above)
Comment (class4kayaker) [inline source/simulator/core.cc]: I was matching the pattern used for updating the field in the core loop. I can see the argument for an Assert being better in this case though.

**Comment by class4kayaker** [comment]
Change made and rebased for nicer history

**Comment by bangerth** [comment]
Ready to merge once the tester is happy.

**Comment by bangerth** [comment]
Ah, but the tester is not happy. There are three failing tests: http://www.math.clemson.edu/~heister/cib/aspect-8.4.1/6f566eb621694332e622aa8b322e5e19466b66f0/log-clang

Can you investigate why these tests now fail?

**Comment by class4kayaker** [comment]
Looking into it. All three are discontinuous composition cases, and looking at the output of a run shows that there was an attempt to access a non-existent matrix entry. I am going to generate and add a discontinuous composition test case for help debugging, and given that this error occured I will also be including it for general testing. My guess is that I am somehow causing the discontinuous case to not allocate the matrix.

**Comment by class4kayaker** [comment]
Ready for another review. The correction has been rebased for neater history.

**Comment by bangerth** [comment]
Can you squash the second and third commits into one?

**Pull request merged**